### PR TITLE
Changes learning_curves to use "step" or "epoch" as x-axis label.

### PR DESCRIPTION
**Pull request by dantreiman**
- Adds frequency and units to training stats
- Adds dataclasses: TrainingResults, TrainingStats, PreprocessedDataset
- training_statistics.json has evaluation frequency added, so visualize learning curves from saved model uses the right units in x axis label

### CODE REVIEW

File: ludwig/api.py [hunk 1]
```diff
@@ -104,6 +106,86 @@
...
+    def __contains__(self, key):
+        return (
+            (key == TRAINING and self.training)
+            or (key == VALIDATION and self.validation)
+            or (key == TEST and self.test)
+        )
+
+    def __getitem__(self, key):
+        # Supports dict-style [] element access for compatibility.
+        return {TRAINING: self.training, VALIDATION: self.validation, TEST: self.test}[key]
```
Comment (justinxzhao) [inline ludwig/api.py:151]: Should we add a Todo here to get rid of these auxiliary/non-standard accessor methods?

**Review by justinxzhao — APPROVED** [review]

**Review by justinxzhao — APPROVED** [review]

File: ludwig/api.py [hunk 2]
```diff
@@ -104,6 +106,88 @@
 logger = logging.getLogger(__name__)
 
 
+@dataclass
```
Comment (connor-mccorm) [inline ludwig/api.py:109]: Should we maybe move these dataclasses into an auxilary file and import them to maintain the api file flow? I feel like these data classes could be pretty easily compartmentalized into a dedicated file which may help with the mental model associated with the api file. WDYT?

**Review by connor-mccorm — COMMENTED** [review]
This is good stuff, had one comment on whether we should be putting the new data classes in the api file.

File: ludwig/api.py (hunk 2 above)
Comment (dantreiman) [reply ludwig/api.py:109]: I put them here because they are used as data types in method of api.py, and I couldn't think of a better idea.

On that note, Arnav and I have discussed further refactoring of api.py this sprint, perhaps defining LudwigModel in `ludwig_model.py` and pulling it into api.py.  So maybe we can separate that out into a follow-up PR.
